### PR TITLE
Port scanner.cc to C

### LINKED ISSUE #84: Migrate scanner.cc to C
Hey, we are thinking about bundling this parser into Neovim for v0.9, however one of the concerns is raising our build requirements to include C++. (https://github.com/neovim/neovim/pull/22458)

I've had a play with the markdown scanner.cc and I think it will be fairly straightforward to port this to C. To replace `vector` I think we could just include https://github.com/neovim/neovim/blob/master/src/klib/kvec.h.

Would this be something you would be open too?

### PR DESCRIPTION
**Pull request by lewis6991**
Resolves #84

### CODE REVIEW

**Comment by lewis6991** [comment]
Ok, this is getting close, got a handful of failures on the markdown parser.

**Comment by lewis6991** [comment]
All passing now.

**Comment by MDeiml** [comment]
Hm, CI is still getting some errors.

**Comment by lewis6991** [comment]
The inline parser has some errors, but those errors where also on head so not sure what's going on. I'll look into it.

Can you adjust permissions so my changes trigger ci without approval?

File: tree-sitter-markdown/src/scanner.c [hunk 1]
```diff
@@ -0,0 +1,1446 @@
+#include <tree_sitter/parser.h>
+#include <wchar.h>
```
Comment (ii14) [inline tree-sitter-markdown/src/scanner.c:2]:
```suggestion
#include <wchar.h>
#include <wctype.h>
#include <string.h>
```

string.h for memcpy, wctype.h for iswlower, iswalpha etc

File: tree-sitter-markdown/src/scanner.c [hunk 2]
```diff
@@ -0,0 +1,1446 @@
...
+// Returns the indentation level which lines of a list item should have at minimum. Should only be
+// called with blocks for which `is_list_item` returns true.
+static uint8_t list_item_indentation(Block block) {
+    return block - LIST_ITEM + 2;
+}
+
+const size_t NUM_HTML_TAG_NAMES_RULE_1 = 3;
+const char* HTML_TAG_NAMES_RULE_1[NUM_HTML_TAG_NAMES_RULE_1] = { "pre", "script", "style" };
+const size_t NUM_HTML_TAG_NAMES_RULE_7 = 62;
+const char* HTML_TAG_NAMES_RULE_7[NUM_HTML_TAG_NAMES_RULE_7] = {
```
Comment (ii14) [inline tree-sitter-markdown/src/scanner.c]: I don't think defining array sizes with `const size_t` is compliant with the C standard. At the same time, I think treesitter expects `NUM_HTML_TAG_NAMES_RULE_1` and `NUM_HTML_TAG_NAMES_RULE_7` symbols to be actually there? So either an additional macro to initialize both, or inline value for the array size

**Comment by MDeiml** [comment]
LGTM, just needs updates to some config files so bindings don't break: https://github.com/MDeiml/tree-sitter-markdown/search?q=scanner.cc

(Sorry I didn't notice that earlier :sweat_smile:)

**Comment by MDeiml** [comment]
Great, thank you.

**Comment by clason** [comment]
New release? ;)

**Pull request merged**